### PR TITLE
Refactoring multi-token

### PR DESCRIPTION
**Pull request by MedovTimur**

### CODE REVIEW

File: contracts/multi-token/src/lib.rs [hunk 1]
```diff
@@ -132,14 +171,178 @@ impl SimpleMTKCore for SimpleMTK {
...
+            .enumerate()
+            .for_each(|(i, id)| self.burn_impl(id, amounts[i]));
+
+        for (id, amount) in ids.iter().zip(amounts.iter()) {
+            let quantity = self.supply.get_mut(id).unwrap();
+            *quantity = quantity.saturating_sub(*amount);
+        }
+
+        Ok(MtkEvent::Transfer {
+            from: msg::source(),
```
Comment (LouiseMedova) [inline contracts/multi-token/src/lib.rs:201]: `msg::source()` is called several times (also in `burn_impl`)
it's necessary to save `msg::source()` value into variable

File: contracts/multi-token/src/lib.rs [hunk 2]
```diff
@@ -132,14 +171,178 @@ impl SimpleMTKCore for SimpleMTK {
...
+            to: ZERO_ID,
```
Comment (LouiseMedova) [inline contracts/multi-token/src/lib.rs:202]: `ActorId::zero()` instead of `ZERO_ID`

File: contracts/multi-token/src/lib.rs [hunk 3]
```diff
@@ -132,14 +171,178 @@ impl SimpleMTKCore for SimpleMTK {
...
+    /// * `to` must be a non-zero account
+    /// Arguments:
+    /// * `to`: An account that will be approved to manage the tokens
+    fn approve(&mut self, to: &ActorId) -> Result<MtkEvent, MtkError> {
+        if to == &ZERO_ID {
+            return Err(MtkError::ZeroAddress);
+        }
+        self.tokens
+            .approvals
+            .entry(msg::source())
```
Comment (LouiseMedova) [inline contracts/multi-token/src/lib.rs:317]: again calling `msg::source()` several times

File: contracts/multi-token/src/lib.rs [hunk 4]
```diff
@@ -132,14 +171,178 @@ impl SimpleMTKCore for SimpleMTK {
...
+    /// Requirements:
+    /// * Only the token owner can call that action
+    /// * `to` must be a non-zero account
+    /// Arguments:
+    /// * `to`: An account that won't be able to manage the tokens
+    fn revoke_approval(&mut self, to: &ActorId) -> Result<MtkEvent, MtkError> {
+        self.tokens
+            .approvals
+            .get_mut(&msg::source())
+            .expect("Caller has not approved any accounts")
```
Comment (LouiseMedova) [inline contracts/multi-token/src/lib.rs:339]: return `Err` instead of panic

File: contracts/multi-token/src/lib.rs [hunk 5]
```diff
@@ -154,91 +357,143 @@ impl SimpleMTKCore for SimpleMTK {
...
-                ids.push(id);
-            }
+                .zip(burn_info.nfts_ids.iter())
+                .try_for_each(|(meta, &id)| {
+                    self.mint_impl(&burn_info.account, &id, NFT_COUNT, meta)
+                })?;
+
+            ids.extend_from_slice(&burn_info.nfts_ids);
         }
+        let quantity = self.supply.get_mut(&id).unwrap();
```
Comment (LouiseMedova) [inline contracts/multi-token/src/lib.rs:400]: don't use `unwrap()`
you should return `Err` instead

**Review by LouiseMedova — CHANGES_REQUESTED** [review]

File: contracts/concert/src/lib.rs [hunk 6]
```diff
@@ -125,108 +136,102 @@ impl Concert {
...
             .expect("CONCERT: Error burning balances");
         }
 
         for actor in &self.buyers {
-            let mut ids = vec![];
-            let mut amounts = vec![];
-            let mut meta = vec![];
             let actor_metadata = self.metadata.get(actor);
             if let Some(actor_md) = actor_metadata.cloned() {
+                let mut ids = vec![];
```
Comment (shamilsan) [inline contracts/concert/src/lib.rs:207]: It is better to set capacity if known.
```suggestion
                let mut ids = Vec::with_capacity(actor_md.len());
```

File: contracts/concert/src/lib.rs [hunk 7]
```diff
@@ -254,13 +259,14 @@ impl From<Concert> for State {
...
         } = value;
 
-        let buyers = buyers.iter().copied().collect();
+        let buyers = buyers.into_iter().collect();
 
         let metadata = metadata
-            .iter()
-            .map(|(k, v)| (*k, v.iter().map(|(k, v)| (*k, v.clone())).collect()))
+            .into_iter()
+            .map(|(k, v)| (k, v.into_iter().map(|(k, v)| (k, v)).collect()))
```
Comment (shamilsan) [inline contracts/concert/src/lib.rs:269]: No need use map here
```suggestion
            .map(|(k, v)| (k, v.into_iter().collect()))
```

File: contracts/multi-token/tests/utils.rs [hunk 8]
```diff
@@ -259,24 +276,55 @@ pub fn transform_internal(
...
+        from: from.into(),
+        to: to.into(),
+    })
+    .encode();
+
+    if let Some(error) = error {
+        assert!(res.contains(&(from, Err::<MtkEvent, MtkError>(error).encode())));
+    } else {
+        assert!(res.contains(&(from, codec)));
+    }
```
Comment (shamilsan) [inline contracts/multi-token/tests/utils.rs:325]: I'd rewrite a little bit.

```suggestion
    let expected = if let Some(error) = error {
        Err(error)
    } else {
        Ok(MtkEvent::RevokeApproval {
            from: from.into(),
            to: to.into(),
        })
    };
    assert!(res.contains(&(from, expected.encode())));
```

**Review by shamilsan — APPROVED** [review]

**Review by LouiseMedova — APPROVED** [review]

**Pull request merged**